File: phenotype-skills/bindings/python/src/phenotype_skills/dependency_resolver.py
"""Dependency resolver for skill graphs."""

from typing import Dict, List, Set, Iterable
from collections import defaultdict

from .types import SkillId
from .skill import Skill


class DependencyResolver:
    """Dependency resolver for skill graphs."""

    def resolve(self, skills: Iterable[Skill]) -> List[Skill]:
        """Resolve all dependencies for a set of skills.
        
        Returns skills in dependency order (dependencies first).
        """
        skill_list = list(skills)
        if not skill_list:
            return []

        # Build dependency graph
        graph: Dict[SkillId, List[SkillId]] = defaultdict(list)
        skill_map = {s.id: s for s in skill_list}

        for skill in skill_list:
            for dep in skill.manifest.dependencies:
                # Find skill that provides this dependency
                provider = next((s for s in skill_list if s.id.value == dep.name), None)
                if provider:
                    graph[skill.id].append(provider.id)

        # Topological sort
        result: List[Skill] = []
        visited: Set[SkillId] = set()
        temp: Set[SkillId] = set()

        def visit(skill_id: SkillId) -> None:
            if skill_id in temp:
                raise ValueError(f"Circular dependency detected involving: {skill_id}")
            if skill_id in visited:
                return

            temp.add(skill_id)
            for dep_id in graph.get(skill_id, []):
                visit(dep_id)
            temp.remove(skill_id)
            visited.add(skill_id)
            result.append(skill_map[skill_id])

        for skill in skill_list:
            visit(skill.id)

        # Reverse to get dependency order (dependencies before dependents)
        result.reverse()
        return result

    def check_circular(self, skills: Iterable[Skill]) -> None:
        """Check for circular dependencies."""
        # Just run resolve - it will raise on circular dependencies
        self.resolve(skills)

    def get_all_dependencies(self, skill: Skill, all_skills: Iterable[Skill]) -> List[Skill]:
        """Get all dependencies for a skill (recursively)."""
        all_skills_list = list(all_skills)
        result: List[Skill] = []
        visited: Set[str] = set()

        def collect_deps(s: Skill) -> None:
            for dep in s.manifest.dependencies:
                if dep.name in visited:
                    continue
                dep_skill = next((x for x in all_skills_list if x.id.value == dep.name), None)
                if dep_skill:
                    visited.add(dep.name)
                    result.append(dep_skill)
                    collect_deps(dep_skill)

        collect_deps(skill)
        return result

    def to_dot(self, skills: Iterable[Skill]) -> str:
        """Generate a DOT graph representation."""
        lines = ["digraph dependencies {", "  rankdir=BT;"]
        skill_list = list(skills)

        # Add nodes
        for skill in skill_list:
            lines.append(f'  "{skill.id}" [label="{skill.id} v{skill.version}"];')

        # Add edges
        for skill in skill_list:
            for dep in skill.manifest.dependencies:
                lines.append(f'  "{dep.name}" -> "{skill.id}";')

        lines.append("}")
        return "\n".join(lines)

File: phenotype-skills/bindings/typescript/src/dependency-resolver.ts
/**
 * Dependency resolver for skill graphs
 */

import { Skill } from './skill';

/**
 * Dependency resolver for skill graphs
 */
export class DependencyResolver {
  /**
   * Resolve all dependencies for a set of skills
   * @returns Skills in dependency order (dependencies first)
   */
  resolve(skills: Skill[]): Skill[] {
    if (skills.length === 0) return [];

    // Build dependency graph
    const graph = new Map<string, string[]>();
    const skillMap = new Map<string, Skill>();

    for (const skill of skills) {
      const id = skill.id.toString();
      skillMap.set(id, skill);
      const deps: string[] = [];
      for (const dep of skill.manifest.dependencies) {
        // Find skill that provides this dependency
        const provider = skills.find((s) => s.id.value === dep.name);
        if (provider) {
          deps.push(provider.id.toString());
        }
      }
      graph.set(id, deps);
    }

    // Topological sort
    const result: Skill[] = [];
    const visited = new Set<string>();
    const temp = new Set<string>();

    const visit = (id: string) => {
      if (temp.has(id)) {
        throw new Error(`Circular dependency detected involving: ${id}`);
      }
      if (visited.has(id)) return;

      temp.add(id);
      const deps = graph.get(id) ?? [];
      for (const depId of deps) {
        visit(depId);
      }
      temp.delete(id);
      visited.add(id);
      const skill = skillMap.get(id);
      if (skill) result.push(skill);
    };

    for (const skill of skills) {
      visit(skill.id.toString());
    }

    // Reverse to get dependency order (dependencies before dependents)
    result.reverse();
    return result;
  }

  /**
   * Check for circular dependencies
   * @throws Error if circular dependencies are detected
   */
  checkCircular(skills: Skill[]): void {
    // Just run resolve - it will throw on circular dependencies
    this.resolve(skills);
  }

  /**
   * Get all dependencies for a skill (recursively)
   */
  getAllDependencies(skill: Skill, allSkills: Skill[]): Skill[] {
    const result: Skill[] = [];
    const visited = new Set<string>();

    const collectDeps = (s: Skill) => {
      for (const dep of s.manifest.dependencies) {
        if (visited.has(dep.name)) continue;
        const depSkill = allSkills.find((x) => x.id.value === dep.name);
        if (depSkill) {
          visited.add(dep.name);
          result.push(depSkill);
          collectDeps(depSkill);
        }
      }
    };

    collectDeps(skill);
    return result;
  }

  /**
   * Generate a DOT graph representation
   */
  toDot(skills: Skill[]): string {
    const lines = ['digraph dependencies {', '  rankdir=BT;'];

    // Add nodes
    for (const skill of skills) {
      lines.push(`  "${skill.id}" [label="${skill.id} v${skill.version}"];`);
    }

    // Add edges
    for (const skill of skills) {
      for (const dep of skill.manifest.dependencies) {
        lines.push(`  "${dep.name}" -> "${skill.id}";`);
      }
    }

    lines.push('}');
    return lines.join('\n');
  }
}
